lib/src/command/add.dart

```diff
@@ -0,0 +1,237 @@
+  String get path => argResults['path'];
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds packages to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final hasGitOptions = gitUrl != null || gitRef != null || gitPath != null;
+    final hasHostOptions = hostUrl != null;
+
+    if ((path != null && hasGitOptions) ||
+        (hasGitOptions && hasHostOptions) ||
+        (hasHostOptions && path != null)) {
+      usageException('Packages must either be a git, hosted, or path package.');
+    }
+
+    final package = _parsePackage(argResults.rest.first);
+
+    /// Perform version resolution in-memory.
+    final updatedPubSpec =
+        await _addPackagesToPubspec(entrypoint.root.pubspec, package);
+    final solveResult = await resolveVersions(
+        SolveType.GET, cache, Package.inMemory(updatedPubSpec));
+
+    final resultPackage =
+        await _getAndAssertResultPackage(solveResult, package);
+
+    /// Update the pubspec.
+    _updatePubspec(resultPackage, package);
+
+    /// Run pub get once we have successfully updated the pubspec
+    await runner.run(['get']);
+  }
+
+  /// Creates a new in-memory [Pubspec] by adding [package] to [original].
+  Future<Pubspec> _addPackagesToPubspec(
+      Pubspec original, PackageInfo package) async {
+    ArgumentError.checkNotNull(original, 'original');
+    ArgumentError.checkNotNull(package, 'package');
+
+    final dependencies = [...original.dependencies.values];
+    var devDependencies = [...original.devDependencies.values];
+
+    if (isDevelopment) {
+      final dependencyNames = dependencies.map((dependency) => dependency.name);
+
+      if (dependencyNames.contains(package.name)) {
+        log.message('${package.name} is already in dependencies. '
+            'Please remove existing entry before adding it to dev_dependencies');
+        await flushThenExit(exit_codes.SUCCESS);
+      }
+
+      devDependencies.add(package.toPackageRange(cache));
+    } else {
+      final devDependencyNames =
+          devDependencies.map((devDependency) => devDependency.name);
+
+      if (devDependencyNames.contains(package.name)) {
+        log.message('${package.name} was found in dev_dependencies. '
+            'Removing ${package.name} and adding it to dependencies instead.');
+        devDependencies =
+            devDependencies.takeWhile((d) => d.name != package.name).toList();
```

Comment:
This will remove any dev dependencies after package.

```suggestion
            devDependencies.where((d) => d.name != package.name).toList();
```